Notebook 01 focused on exploratory data auditing and dataset understanding. The objective was to analyze the dataset structure, identify the target variable, inspect missing values, detect duplicate records, understand feature types, and assess overall data quality before proceeding to preprocessing and feature engineering.

In [8]:
# ==========================================
# MULE ACCOUNT DETECTION
# PHASE 1 : DATA AUDIT
# ==========================================

import pandas as pd
import numpy as np
import os

# ------------------------------------------
# CHECK CURRENT DIRECTORY
# ------------------------------------------

print("=" * 60)
print("CURRENT WORKING DIRECTORY")
print("=" * 60)

print(os.getcwd())

# ------------------------------------------
# LOAD DATASET
# ------------------------------------------

df = pd.read_csv("../data/raw/DataSet.csv")

print("\n" + "=" * 60)
print("DATASET LOADED SUCCESSFULLY")
print("=" * 60)

# ------------------------------------------
# DATASET SHAPE
# ------------------------------------------

print("\nDATASET SHAPE")
print("-" * 40)

print("Rows    :", df.shape[0])
print("Columns :", df.shape[1])

# ------------------------------------------
# FIRST 5 ROWS
# ------------------------------------------

print("\nFIRST 5 ROWS")
print("-" * 40)

print(df.head())

# ------------------------------------------
# DATASET INFO
# ------------------------------------------

print("\nDATASET INFO")
print("-" * 40)

df.info()

# ------------------------------------------
# DATA TYPES COUNT
# ------------------------------------------

print("\nDATA TYPES COUNT")
print("-" * 40)

print(df.dtypes.value_counts())

# ------------------------------------------
# NUMERICAL FEATURES
# ------------------------------------------

numerical_cols = df.select_dtypes(
    include=["int64", "float64"]
).columns

print("\nNUMERICAL FEATURES")
print("-" * 40)

print("Total Numerical Features :", len(numerical_cols))

# ------------------------------------------
# CATEGORICAL FEATURES
# ------------------------------------------

categorical_cols = df.select_dtypes(
    include=["object"]
).columns

print("\nCATEGORICAL FEATURES")
print("-" * 40)

print("Total Categorical Features :", len(categorical_cols))

print("\nCategorical Columns:")

for col in categorical_cols:
    print(col)

# ------------------------------------------
# MISSING VALUE ANALYSIS
# ------------------------------------------

print("\nMISSING VALUE ANALYSIS")
print("-" * 40)

missing_report = pd.DataFrame({
    "Column": df.columns,
    "Missing_Count": df.isnull().sum().values,
    "Missing_Percentage":
        (df.isnull().sum().values / len(df)) * 100
})

missing_report = missing_report.sort_values(
    by="Missing_Percentage",
    ascending=False
)

print(missing_report.head(20))

# ------------------------------------------
# COLUMNS > 80% MISSING
# ------------------------------------------

high_missing = missing_report[
    missing_report["Missing_Percentage"] > 80
]

print("\nCOLUMNS WITH >80% MISSING VALUES")
print("-" * 40)

print("Count :", len(high_missing))

print(high_missing.head(20))

# ------------------------------------------
# DUPLICATE RECORDS
# ------------------------------------------

duplicates = df.duplicated().sum()

print("\nDUPLICATE RECORDS")
print("-" * 40)

print("Duplicate Rows :", duplicates)

# ------------------------------------------
# TARGET ANALYSIS
# ------------------------------------------

target_col = "F3924"

print("\nTARGET COLUMN")
print("-" * 40)

print("Target :", target_col)

print("\nTARGET VALUE COUNTS")
print("-" * 40)

print(df[target_col].value_counts())

print("\nTARGET DISTRIBUTION (%)")
print("-" * 40)

print(
    round(
        df[target_col].value_counts(normalize=True) * 100,
        2
    )
)

# ------------------------------------------
# NUMERICAL SUMMARY
# ------------------------------------------

print("\nNUMERICAL SUMMARY")
print("-" * 40)

print(df.describe())

# ------------------------------------------
# CATEGORICAL FEATURE DETAILS
# ------------------------------------------

print("\nCATEGORICAL FEATURE DETAILS")
print("-" * 40)

for col in categorical_cols:

    print(f"\nColumn : {col}")

    print(
        "Unique Values :",
        df[col].nunique()
    )

    print(
        df[col]
        .value_counts()
        .head()
    )

# ------------------------------------------
# SAVE REPORT
# ------------------------------------------

os.makedirs("../reports", exist_ok=True)

missing_report.to_csv(
    "../reports/missing_value_report.csv",
    index=False
)

print("\nMissing Value Report Saved")

# ------------------------------------------
# FINAL SUMMARY
# ------------------------------------------

print("\n" + "=" * 60)
print("FINAL DATASET SUMMARY")
print("=" * 60)

print("Rows                     :", df.shape[0])
print("Columns                  :", df.shape[1])
print("Numerical Features       :", len(numerical_cols))
print("Categorical Features     :", len(categorical_cols))
print("Duplicate Rows           :", duplicates)
print("Target Column            :", target_col)
print("Columns >80% Missing     :", len(high_missing))

print("=" * 60)

CURRENT WORKING DIRECTORY
/Users/avirajmurkute/Desktop/Mule-Account-Detection/notebooks

DATASET LOADED SUCCESSFULLY

DATASET SHAPE
----------------------------------------
Rows    : 9082
Columns : 3925

FIRST 5 ROWS
----------------------------------------
   Unnamed: 0  F1  F2  F3  F4  F5  F6  F7  F8  F9  ...  F3915  F3916  F3917  \
0           1 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...      0      1      0   
1           2 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...      0      1      0   
2           3 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...      0      1      0   
3           4 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...      0      1      0   
4           5 NaN NaN NaN NaN NaN NaN NaN NaN NaN  ...      0      1      0   

   F3918  F3919  F3920  F3921  F3922  F3923  F3924  
0      0      1      0      0      1      0      0  
1      0      1      0      1      0      0      0  
2      0      1      0      0      1      0      0  
3      0      1      0      0      1      0      0  
4 

/var/folders/9x/2_977c9s79q4_3v2tqykf1t00000gn/T/ipykernel_11052/2665231879.py:84: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(



DUPLICATE RECORDS
----------------------------------------
Duplicate Rows : 0

TARGET COLUMN
----------------------------------------
Target : F3924

TARGET VALUE COUNTS
----------------------------------------
F3924
0    9001
1      81
Name: count, dtype: int64

TARGET DISTRIBUTION (%)
----------------------------------------
F3924
0    99.11
1     0.89
Name: proportion, dtype: float64

NUMERICAL SUMMARY
----------------------------------------
        Unnamed: 0           F1           F2           F3           F4  \
count  9082.000000  2591.000000  1532.000000  1448.000000  2591.000000   
mean   4541.500000     0.679722     0.656129     0.724731     0.691721   
std    2621.891906     0.392573     0.398420     0.408869     0.401105   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%    2271.250000     0.430000     0.330000     0.500000     0.380000   
50%    4541.500000     1.000000     1.000000     1.000000     1.000000   
75%    6811.750000     1.000000

Summary of Notebook 01: Data Audit

Purpose: Before building the mule account detection model, we first inspect and understand the dataset.

What we are doing
Load the Dataset
Read the CSV file into Pandas.
Make the data available for analysis.
Check Dataset Size
Find number of rows (accounts).
Find number of columns (features).
Preview the Data
Look at the first few records.
Understand what kind of information is present.
Identify Data Types
Find numerical columns.
Find categorical (text) columns.
Know which features need encoding later.
Analyze Missing Values
Check which features contain missing data.
Identify columns with very high missing percentages.
Check Duplicate Records
Find duplicate rows that may affect model performance.
Analyze the Target Variable (F3924)
Count mule vs legitimate accounts.
Check whether the dataset is balanced or imbalanced.
Generate a Dataset Report
Save information about missing values and dataset quality for future preprocessing.
Outcome of Notebook 01

After completing this notebook, we will know:

Total number of accounts.
Total number of features.
Numerical and categorical features.
Features with many missing values.
Presence of duplicate records.
Distribution of mule and legitimate accounts.
Overall dataset quality.
Why this is important

Before training any ML model, we must understand the data. This notebook helps us decide:

Keep which features?
Remove which features?
Handle missing values how?
Need encoding or not?
Is the target balanced?

Only after answering these questions can we move to the next phase:

Notebook 02
Data Cleaning & Missing Value Handling

Think of Notebook 01 as a medical checkup of the dataset before starting treatment (machine learning).

You can directly paste this into your project report/notebook.

---

## Data Audit Summary

### Objective

The objective of the Data Audit phase was to understand the structure and quality of the dataset before performing preprocessing, feature engineering, and machine learning model development for mule account detection.

---

### Dataset Overview

The dataset contains:

* **Rows (Accounts):** 9,082
* **Columns (Features):** 3,925

Each row represents a bank account and each column represents a feature describing account behavior, transaction patterns, customer information, or derived banking indicators.

The target variable for classification is:

```text
F3924
```

where the model will predict whether an account is suspicious (mule account) or legitimate.

---

### Target Variable Analysis

The distribution of the target variable is:

| Class                       | Count |
| --------------------------- | ----- |
| 0 (Legitimate Account)      | 9001  |
| 1 (Suspicious/Mule Account) | 81    |

Percentage distribution:

| Class      | Percentage |
| ---------- | ---------- |
| Legitimate | 99.11%     |
| Suspicious | 0.89%      |

#### Observation

The dataset is highly imbalanced because suspicious accounts represent less than 1% of the total records. This is a common characteristic of real-world fraud detection datasets and will require special handling during model development.

---

### Duplicate Record Analysis

The dataset contains:

```text
Duplicate Rows = 0
```

#### Observation

No duplicate records were found in the dataset. Therefore, no duplicate removal is required during preprocessing.

---

### Missing Value Analysis

The analysis identified:

```text
Columns with more than 80% missing values = 908
```

#### Observation

A significant number of features contain a large proportion of missing values. These columns may contribute little useful information and will be evaluated for removal during the data cleaning phase.

---

### Initial Feature Inspection

The dataset contains a column named:

```text
Unnamed: 0
```

#### Observation

This column appears to be an index column generated during dataset export and is not expected to contribute to mule account prediction. It will likely be removed during preprocessing.

---

### Key Findings

The Data Audit phase revealed the following important insights:

1. The dataset contains 9,082 records and 3,925 features.
2. The target variable is F3924.
3. The dataset is highly imbalanced with only 81 suspicious accounts.
4. No duplicate records were found.
5. A total of 908 features contain more than 80% missing values.
6. The dataset contains a large number of features, making feature selection and dimensionality reduction important steps in the pipeline.
7. The presence of significant missing data indicates that data cleaning and preprocessing will be critical before model training.

---

### Next Step

The next phase of the project will focus on:

* Removing unnecessary columns.
* Handling missing values.
* Identifying numerical and categorical features.
* Preparing the dataset for feature selection and feature engineering.
* Creating a cleaned dataset for machine learning model development.

This phase will form the foundation for discovering the most important indicators of suspicious and mule account behavior.
